### ISSUE TITLE
Proposal: update `:where(:root,  {component}), ` -> ~~`:where(:root), [data-reset={component}]`~~ `:root, pf-{component}-reset]`

### ISSUE THREAD

**Issue by mattnolting** [patternfly/patternfly#6647]
# TLDR:

`:where(:root), :where(.selector)` is a redundancy and produces the same result as `:root, .selector`. We created this model to address scoping issues. For example, `page` defines `menu-toggle`'s `background-color` as blue, but card (within page) requires `menu-toggle` to use default colors. You would then have to redefine the `menu-toggle` color in `card`. The purpose of `:where(:root), :where(.selector)` is to prevent this scoping issue, but unfortunately it presents a new issue. It prevents `page` from defining `menu-toggle` background color entirely without increasing specificity. 

This results in having to write something like:
```
.page {
  .menu-toggle {
    --menu-toggle--BackgroundColor: blue;
  }
}
```

Instead of:
```
.page {
  --menu-toggle--BackgroundColor: blue;
}
```

Additionally, it means that users can't have a flat theming file:

**Instead of this:**

```
.my-custom-theme {
  --menu-toggle--BackgroundColor: blue;
}
```

**Users must also increase specificity** 

```
.my-custom-theme {
  .menu-toggle {
    --menu-toggle--BackgroundColor: blue;
  }
}
```

**The solution?**
* Components need a reset that is independent of their base class. 
* Both `:root` and `reset` require specificity of 10

**Possible names**
* `:where(:root), [data-reset="selector"]` @mcoker @srambach update
* `:root, data-reset-menu-toggle {}`
* `:root, [data-reset-style="menu-toggle"] {}

## Additional benefits of this method

* **Flexibility:** Using this pattern `[data-reset-style="component"]` is ubiquitous. It could be included in any component, not just  the specific component. So, `.card` itself can reset `menu-toggle` or any other component for example.
* **Nestability:** `.page` can define `.menu-toggle`, `.page-section` can reset `.menu-toggle`, `.card` can reset `.menu-toggle` to its default state.  

**Example**

**CSS**
```
.page {
  --menu-toggle--BackgroundColor: blue;
}

.page-section {
  --menu-toggle--BackgroundColor: red;
}
```

**HTML**
```
<div class="page">
  <button class="menu-toggle"></button> // color blue
  <div class="page-section">
    <button class="menu-toggle"></button> // color red
    <div class="card" data-reset="menu-toggle">
      <button class="menu-toggle"></button> // reset to default state
      // here menu-toggle can be redefined by parent
        // here menu-toggle reset to default
    </div>
  </div>
</div>
```

**My custom theme**
```
.my-theme {
  --menu-toggle--BackgroundColor: fuscia; // this can then be reset by [data-reset="menu-toggle"]
}
```

## Process and approach

* Replace `:where(:root), :where(.component) {}` in a 3-5 components of varying complexity. 
* Run visual regression tests
* Analyze results and triage feasibility 

## Additional thoughts

**I haven't tested this, but we may be able to do something like this with card CSS**

```
:root,
[data-reset="card"],
[data-reset="menu-toggle"] {
  
}
```

**Comment by mattnolting** [comment]
Notes from 5/14 discussion

**Q:** Why should we remove `:where()`
**A:** It may be an anti-pattern, but we will look into this. The more important aspect of this proposal is the `.selector` prevents the cascade from behaving as expected. This approach to scoping prevents parent elements from styling child component without increasing specificity.

**Comment by mattnolting** [comment]
# Notes from 5/15

## Potential issues

`[data-reset="selector"]` may be expensive. Explore this. 
Useful links:
* https://devtoolstips.org/tips/en/find-expensive-selectors/
* https://csswizardry.com/2011/09/writing-efficient-css-selectors/

## Possible alternative
  * possible alternatives - use a class instead
    * what is the syntax? `pf-v6-card-reset`

**Comment by garrett** [comment]
I wonder if something like `[data-reset="selector"].pf-v6-reset` could work? Going from right to left with [bottom-up parsing](https://en.wikipedia.org/wiki/Bottom-up_parsing) ([like CSS selector matching does](https://stackoverflow.com/questions/5797014/why-do-browsers-match-css-selectors-from-right-to-left/5813672#5813672)), it'd look for everything with the class first, _then_ do the more expensive selecting on that subset, instead of trying to do an attribute selector across the entire DOM.

The side effect is that you'd need to specify it in two places instead. I wonder if there could be a more general reset and then you could specify? That could probably be targeted like this then, if so?

```css
:not([data-reset]).pf-v6-reset { /* reset all here */ }
```

...but this is all off the top of my head and not tested, after a day full of meetings. :wink: 

But I'm not sure if this matters or not, as the right-to-left thing might be just for combinations making up a selector and not necessarily compound selectors like this. (I'm not sure if this would be considered a combination or compound, however.)

**Comment by garrett** [comment]
Here's another blog post about the CSS selector matching functionality that's in Edge (and soon in Chrome), from the Edge dev team, which walks through how to use it with some examples too:

https://blogs.windows.com/msedgedev/2023/01/17/the-truth-about-css-selector-performance/
